### PR TITLE
[WIP] Fix timer race condition

### PR DESCRIPTION
**Pull request by jay-zhuang**

### CODE REVIEW

File: util/timer.h [hunk 1]
```diff
@@ -121,17 +123,20 @@ class Timer {
   bool Shutdown() {
     {
       InstrumentedMutexLock l(&mutex_);
+      shutting_down_ = true;
```
Comment (mrambacher) [inline util/timer.h:126]: Perhaps I do not understand some of the cases enough but won't this cause the Start task to wait forever if running_ = false?  Right below here it returns false "if (!running_)", without resetting shutdown or signaling any conditions.

**Review by mrambacher — COMMENTED** [review]
I think #7722 was also proposed as a fix to the same problem.  I do not know which one is correct or better but wanted to bring it to attention.

**Comment by nikitamikhaylov** [comment]
> I think #7722 was also proposed as a fix to the same problem. I do not know which one is correct or better but wanted to bring it to attention.

I thought that I was easy to fix, but no :)

File: util/timer.h (hunk 1 above)
Comment (jay-zhuang) [reply util/timer.h:126]: Nice catch. Thanks @mrambacher

**Pull request closed without merging**